File: go/11/main.go
package main

import (
	"fmt"
	"reflect"
)

func main() {
	assertEq(49, maxArea([]int{1, 8, 6, 2, 5, 4, 8, 3, 7}))
	assertEq(1, maxArea([]int{1, 1}))
}

func maxArea(height []int) int {
	var res int
	i := 0
	j := len(height) - 1
	for i < j {
		res = max(res, (j-i)*min(height[i], height[j]))
		if height[i] < height[j] {
			i++
		} else {
			j--
		}
	}

	return res
}

func assertEq(a, b any) {
	if reflect.DeepEqual(a, b) {
		fmt.Printf("Ok: %v\n", a)
	} else {
		fmt.Printf("Failed: %v != %v\n", a, b)
	}
}

File: python/11/main.py
from typing import *

def maxArea(height: List[int]) -> int:
    res = 0
    i, j = 0, len(height)-1
    while i < j:
        w = j - i
        hi = height[i]
        hj = height[j]
        h = min(hi, hj)
        res = max(res, w*h)
        if hi < hj:
            while i < j and height[i] <= hi:
                i += 1
        else:
            while i < j and height[j] <= hj:
                j -= 1
    return res

def assert_eq(a, b):
    if a == b:
      print("Correct")
    else:
      print(f"Wrong: {a} != {b}")

assert_eq(49, maxArea([1,8,6,2,5,4,8,3,7]))
assert_eq(1, maxArea([1,1]))
